src/platforms/esp32/main/component_drivers.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+##
+## Copyright (c) 2020 fred@dushin.net
+## All rights reserved.
+##
+## Licensed under the Apache License, Version 2.0 (the "License");
+## you may not use this file except in compliance with the License.
+## You may obtain a copy of the License at
+##
+##     http://www.apache.org/licenses/LICENSE-2.0
+##
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+##
+
+import os
+import sys
+
+def load(path) :
+    with open(path) as f:
+        return f.read()
+
+def insert(component_driver, component_drivers_txt) :
+
+    declarations_pos = component_drivers_txt.find("// DECLARATIONS")
+    buf = component_drivers_txt[:declarations_pos]
+    buf += "Context *%s_create_driver(GlobalContext *global, term opts);\n" % component_driver
+    buf += "void %s_init(GlobalContext *global);\n" % component_driver
+
+    create_driver_pos = component_drivers_txt.find("// CREATE_DRIVER")
+    buf += component_drivers_txt[declarations_pos:create_driver_pos]
+    buf += "if (strcmp(driver_name, \"%s\") == 0) { return %s_create_driver(global, opts); }\n    " % (driver_name(component_driver), component_driver)
+
+    init_pos = component_drivers_txt.find("// INIT")
+    buf += component_drivers_txt[create_driver_pos:init_pos]
+    buf += "%s_init(global);\n    " % component_driver
+    buf += component_drivers_txt[init_pos:]
+
+    return buf
+
+def driver_name(component_driver) :
+    if component_driver.endswith("_driver") :
+        idx = component_driver.find("_driver")
+        return component_driver[:idx]
+    else :
+        return component_driver
+
+def strip_line(line) :
+    idx = line.find('#')
+    if idx != -1 :
+        line = line[:idx]
+    return line.strip()
+
```

Comment:
this kind of generators always worry me, they are so powerful and they enable us new scenarios, on the other hand they might make things a bit more obscure.
We should write some comments here documenting well what kind of inputs they expect, what kind of output they make and how end users should avoid to directly interact with them (they should instead user better tools such as make menuconfig).

Reply:
Please check the Readme that I added, which explains the internals, and is targeted for developers.